File: data-structure-and-algorithm/basic/single_list/single_list_reverser/SingleListReverser.go
package main

import (
	"fmt"
	"strconv"
)

// ListNode 单链表定义
type ListNode struct {
	val  int
	next *ListNode
}

func (node *ListNode) String() string {
	var next string
	if node.next == nil {
		next = "nil"
	} else {
		next = node.next.String()
	}
	return "[val:" + strconv.Itoa(node.val) + ", next:" + next + "]"
}

func reverse(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	var cur = head.next
	head.next = nil
	var tmp *ListNode
	for cur != nil {
		tmp = cur.next
		cur.next = head
		head = cur
		cur = tmp
	}
	return head
}

func reverseByRecursion(head *ListNode) *ListNode {
	if head == nil || head.next == nil {
		return head
	}
	var pre = reverse(head.next)
	head.next.next = head
	head.next = nil
	return pre
}

func main() {
	node3 := ListNode{
		val:  3,
		next: nil,
	}
	node2 := ListNode{
		val:  2,
		next: &node3,
	}
	node1 := ListNode{
		val:  1,
		next: &node2,
	}
	fmt.Println(reverse(&node1).String())
	fmt.Println(reverseByRecursion(&node3).String())
}

File: data-structure-and-algorithm/basic/single_list/single_list_reverser/SingleListReverser.py
class ListNode:
    def __init__(self, val: int, next):
        self.val = val
        self.next = next

    def __str__(self):
        return f"[val:{self.val}, next:{str(self.next)}]"

def reverse(head: ListNode) -> ListNode:
    if head is None:
        return head
    cur: ListNode = head.next
    head.next = None
    tmp: ListNode = None
    while cur is not None:
        tmp = cur.next
        cur.next = head
        head = cur
        cur = tmp
    return head

def reverse_by_recursion(head: ListNode) -> ListNode:
    if head is None or head.next is None:
        return head
    pre: ListNode = reverse_by_recursion(head.next)
    head.next.next = head
    head.next = None
    return pre

if __name__ == "__main__":
    node3 = ListNode(3, None)
    node2 = ListNode(2, node3)
    node1 = ListNode(1, node2)
    print(reverse(node1))
    print(reverse_by_recursion(node3))

